File: mike/python/icecream.py
n, k = map(int, input().split())
stalls = list(map(int, input().split()))

bottom = ans = 1
top = (stalls[-1] - stalls[0]) // (k - 1)

while bottom <= top:
    curr, count = 0, 1
    mid = (bottom + top) // 2
    for next in range(1, n):
        if stalls[next] - stalls[curr] < mid:
            continue
        curr = next
        count += 1
    if count < k:
        top = mid - 1
    else:
        ans = mid
        bottom = mid + 1

print(ans)

File: mike/cpp/icecream.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    int i, n, k, curr, next, count, bottom, mid, top, ans;
    vector<int> stalls;

    cin >> n >> k;
    stalls.reserve(n);
    for (i = 0; i < n; i++) {
        cin >> curr;
        stalls.push_back(curr);
    }

    bottom = ans = 1;
    // stalls 2 <= n
    top = (stalls.back() - stalls.front()) / (k - 1);

    while (bottom <= top) {
        curr = 0;
        count = 1;
        mid = (bottom + top) / 2;

        for (next = 1; next < n; next++) {
            if (stalls[next] - stalls[curr] < mid) continue;

            curr = next;
            count++;
        }

        if (count < k) {
            top = mid - 1;
        } else {
            ans = mid;
            bottom = mid + 1;
        }
    }

    cout << ans << endl;

    return 0;
}